### PR TITLE
Settings screen with the ability to change your name

### PR DESCRIPTION
**Pull request by clenrock**
- Relevant Issues: #178 
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository a settings screen with a lot of options. The ability to change your name is the only one that really works at the moment.

## Why
This Pull Request is needed because when you have a profile maybe in the future you would like to change it.

## How
This feature can be viewed/tested within the project by uninstalling the app, creating a new profile and then going to the settings screen via MainActivity.

## Alternative implementation
Now I've used the default Android implementation of the preferences, but I first tried to make a screen of my own, which became very ugly.

## Notes
Please give the look and the feel while going through the settings as much attention as the code.
This is an early PR, I will try to fix the errors still in the code today.

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/AppCompatPreferenceActivity.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import android.support.v7.widget.Toolbar;
+import android.view.MenuInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A {@link android.preference.PreferenceActivity} which implements and proxies the necessary calls
+ * to be used with AppCompat.
+ */
+public abstract class AppCompatPreferenceActivity extends PreferenceActivity {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/AppCompatPreferenceActivity.java]: I hate the name of this activity, but unfortunately this is the way to go in Android

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+import java.util.List;
+
+/**
+ * A {@link PreferenceActivity} that presents a set of application settings. This
+ * is the place where your profile is changed.
+ */
+public class SettingsActivity extends AppCompatPreferenceActivity {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("SettingsActivity");
+    private static String DISPLAY_NAME = "display_name";
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Either make this final or change it to camelCase

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    private static IServiceLocator serviceLocator;
+    private static boolean FIRST_LOAD = true;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: same

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+    /**
+     * A preference value change listener that updates the preference's summary
+     * to reflect its new value.
+     */
+    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
+        @Override
+        public boolean onPreferenceChange(Preference preference, Object value) {
+            String stringValue = value.toString();
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: log

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+            if (preference.getKey().equals(DISPLAY_NAME)) {
+                updateDisplayName(preference, stringValue);
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: log

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            } else if (preference instanceof ListPreference) {
+                // For list preferences, look up the correct display value in
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: log

File: app/src/main/res/xml/pref_general.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
+    <CheckBoxPreference android:title="Remember Bluetooth verification method"
```
Comment (jverbraeken) [inline app/src/main/res/xml/pref_general.xml]: enter

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    private static volatile IServiceLocator serviceLocator;
+
+    /**
+     * A preference value change listener that updates the preference's summary
+     * to reflect its new value.
+     */
+    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
+        @Override
+        public boolean onPreferenceChange(final Preference preference, final Object value) {
+            final String stringValue = value.toString();
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: log

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+            if (preference.getKey().equals(DISPLAY_NAME)) {
+                updateDisplayName(preference, stringValue);
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: log

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 2 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Either make this final or change it to camelCase

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: same

**Comment by jverbraeken** [comment]
@clenrock I have no idea why my last two comments are outdated immediately and outdated immediately again after I changed them, but yeah, just that you know that those comments belong to my PR

**Comment by clenrock** [comment]
@jverbraeken I pushed the same exact minute ;)

File: app/src/main/res/values/array.xml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string-array name="verification_methods">
+        <item>Visual verification</item>
+        <item>Rhythm verification</item>
+        <item>Choose when pairing</item>
+    </string-array>
```
Comment (ericcornelissen) [inline app/src/main/res/values/array.xml]: Why did you create a separate resource for this? Also (if you want to keep it a separate resource) give it a better name and do the same for the dutch translation :wink:

File: app/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -49,22 +49,28 @@
...
         <activity
             android:name=".activities.QRExchangeKeyActivity"
             android:theme="@style/AppTheme.NoActionBar" />
-
-        <service android:name=".modules.pairing.AndroidBluetoothService" />
+        <activity
+            android:name=".activities.SettingsActivity"
+            android:label="@string/title_activity_settings" />
+        <activity
+            android:name=".activities.KeyManagementActivity" />
```
Comment (ericcornelissen) [inline app/src/main/AndroidManifest.xml]: Can you keep it so that all the activities are listed before the services?

(Also, why did you change the enters in this file? In particular: L20, L25 and L75)

File: app/src/main/res/values/array.xml (hunk 10 above)
Comment (clenrock) [inline app/src/main/res/values/array.xml]: @ericcornelissen I think it's need to have this file if we would like to expand it with more static arrays. The name 'array.xml' is the default name for arrays, what would you give as a different name?

Comment (ericcornelissen) [inline app/src/main/res/values/array.xml]: > I think it's needed to have this file if we would like to expand it with more static arrays.

In that case, I would suggest you do the same for [the dutch version of `strings.xml`](https://github.com/ericcornelissen/NervousFish/pull/214/files#diff-b0d05aebee027c15c77e36003f7bcadcR108)


> The name 'array.xml' is the default name for arrays, what would you give as a different name?

Seems a bit odd to me given .e.g string**s**.xml, but how does array**s**.xml sound?

File: app/src/main/AndroidManifest.xml (hunk 11 above)
Comment (clenrock) [inline app/src/main/AndroidManifest.xml]: I think that was done automatically because I use "Create Activity" when you press on a package, which automatically adds it to the manifest. I will change it

File: app/src/main/java/com/nervousfish/nervousfish/activities/AAppCompatPreferenceActivity.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+    /**
+     * Returns the {@link AppCompatDelegate}.
+     * @return the {@link AppCompatDelegate}
+     */
+    private AppCompatDelegate getDelegate() {
+        if (mDelegate == null) {
+            mDelegate = AppCompatDelegate.create(this, null);
+        }
+        return mDelegate;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/AAppCompatPreferenceActivity.java]: Why isn't this code in the constructor and don't you use `this.mDelegate` instead of `getDelegate()`?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+                                : "");
+
+
+            } else {
+                LOGGER.info("Preference changed which is not a ListPreference, and not the display name");
+                // For all other preferences, set the summary to the value's
+                // simple string representation.
+                preference.setSummary(stringValue);
+            }
+            return true;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Are you sure you always want to return `true`?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        public boolean onPreferenceChange(final Preference preference, final Object value) {
+            LOGGER.info("Preference changed");
+            final String stringValue = value.toString();
+
+            if (preference.getKey().equals(DISPLAY_NAME)) {
+                LOGGER.info("Preference changed at the display name");
+                updateDisplayName(preference, stringValue);
+            } else if (preference instanceof ListPreference) {
+                LOGGER.info("Preference changed for a ListPreference");
+                // For list preferences, look up the correct display value in
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Can you abstract this into its own implementation similar to [when the display name is changed](https://github.com/ericcornelissen/NervousFish/pull/214/files#diff-bf06f4c02cc485ea11575068ccada50eR53)

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+                        .getString(preference.getKey(), ""));
+    }
+
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setupActionBar();
+
+        final Intent intent = getIntent();
+        if (serviceLocator == null) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Why did you put in this if-statement?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+    /**
+     * This method stops fragment injection in malicious applications.
+     * Make sure to deny any unknown fragments here.
+     *
+     * @param fragmentName The name of the fragment
+     * @return A boolean which is true when the fragment is valid
+     */
+    protected boolean isValidFragment(final String fragmentName) {
+        return PreferenceFragment.class.getName().equals(fragmentName)
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Maybe put `PreferenceFragment.class.getName()` into a local variable?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        public void onCreate(final Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            addPreferencesFromResource(R.xml.pref_general);
+            setHasOptionsMenu(true);
+
+            // Bind the summaries of EditText/List/Dialog/Ringtone preferences
+            // to their values. When their values change, their summaries are
+            // updated to reflect the new value, per the Android Design
+            // guidelines.
+            bindPreferenceSummaryToValue(findPreference("choose_verification_use_every_time"));
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Change "choose_verification_use_every_time" into "choose_verification_method_every_time"

Also, maybe use a enum for this?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        }
+
+        @Override
+        public boolean onOptionsItemSelected(final MenuItem item) {
+            final int id = item.getItemId();
+            if (id == android.R.id.home) {
+                startActivity(new Intent(getActivity(), SettingsActivity.class));
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Why don't you use `startActivity(new Intent(this, SettingsActivity.class));`?

File: app/src/main/res/values-nl/strings.xml [hunk 19]
```diff
@@ -87,4 +87,21 @@
     <string name="show_qr_code">Toon QR code</string>
     <string name="contact_set_name">Geef contact een naam</string>
     <string name="contact_exists_message">Contact bestaat al, geef aub een unieke naam</string>
+
+    <!-- Strings related to Settings -->
+    <string name="title_activity_settings">Settings</string>
+    <string name="content_description_settings_button">Settings knop drie punten</string>
+
+    <!-- Example General settings -->
```
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: Why is this named with "Example"?

File: app/src/main/res/values-nl/strings.xml [hunk 20]
```diff
@@ -87,4 +87,21 @@
...
+    <string name="pref_header_general">Algemeen</string>
+
+    <string name="pref_verification_method_to_use">Verificatie methode om te gebruiken</string>
+
+    <!-- Settings for Profile -->
+    <string name="pref_header_profile">Profiel informatie</string>
+
+    <string name="display_name">Naam</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: What do you think about using "Publieke naam" here?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 13 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Because of extension I have to return a boolean, but I'm not sure if that should be true or false. In the example I used they also returned true after setting the summaries of the elements.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 15 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Because this method is also run when opening a fragment, and it should only be set the first time the SettingsActivity is created.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 16 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: I'm not sure because it is only used once.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 15 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Wierd, but OK 😕

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 13 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: From the documentation ([link](https://developer.android.com/reference/android/preference/Preference.OnPreferenceChangeListener.html)):
> True to update the state of the Preference with the new value.


Would you be OK with putting a separate `return true;` in each branch to make sure you explicitly have to return if you want to set the setting? (although I'm not sure if our static analysis tools are OK with that 😓)

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 16 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Are you sure it is only used once?
```java
protected boolean isValidFragment(final String fragmentName) {
    return PreferenceFragment.class.getName().equals(fragmentName)
        || GeneralPreferenceFragment.class.getName().equals(fragmentName)
        || ProfilePreferenceFragment.class.getName().equals(fragmentName);
}
```

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 18 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: This can't be used because it's in a fragment, not an activity itself.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 16 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Yes

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 13 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: But what to return in the end? False or throw an error? I'm not sure I want to change this

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 18 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Seems a bit odd te me, but sure 😕 Would you mind splitting the code and adding comments? e.g.
```java
final Activity activity = this.getActivity(); // I'm guessing this returns an Activity
final Intent intent = new Intent(activity, SettingsActivity.class); // Needed bacause we're in a Fragment
this.startActivity(intent);
```

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 13 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: I guess throw an error, if we test properly we should always catch that before release... I get your concern but I like being explicit in situations like these...

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 21]
```diff
@@ -219,7 +227,7 @@ public void onCreate(final Bundle savedInstanceState) {
             // to their values. When their values change, their summaries are
             // updated to reflect the new value, per the Android Design
             // guidelines.
-            bindPreferenceSummaryToValue(findPreference("choose_verification_use_every_time"));
+            bindPreferenceSummaryToValue(findPreference("choose_verification_method_every_time"));
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Could you use either a `ConstantKeywords` or a `this.getString(R.strings.id)` here?

File: app/src/main/res/values-nl/strings.xml [hunk 22]
```diff
@@ -97,20 +97,25 @@
...
-    <!-- Settings for Profile -->
+    <!-- Profile Settings -->
     <string name="pref_header_profile">Profiel informatie</string>
 
-    <string name="display_name">Naam</string>
+    <string name="display_name">Publieke naam</string>
 
     <!-- KeyManagementActivity -->
     <string name="pref_key_management_header">Key management</string>
+=======
```
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: Interesting addition 🤔

Comment (clenrock) [inline app/src/main/res/values-nl/strings.xml]: Mmmm...

File: app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java [hunk 23]
```diff
@@ -15,6 +15,8 @@
     public static final String WAIT_MESSAGE = "Wait message";
     public static final String DATA_RECEIVED = "Data received";
     public static final String TAP_DATA = "Tap data";
+    public static final String CHOOSE_VERIFICATION_PREFERENCE = "choose_verification_method_every_time";
+    public static final String DISPLAY_NAME = "display_name";
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java]: Why are you not using the value in `strings.xml` for this?

Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java]: Because it's a "key" not a string

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 13 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Sure, I'll change it

File: app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java (hunk 23 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java]: What do you mean?

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/ConstantKeywords.java]: Because the user will never see that string, it's only used internally because some Android libraries don't support enums for some reason and require you to use Strings to identify things.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 24]
```diff
@@ -18,17 +18,23 @@
...
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import edu.umd.cs.findbugs.annotations.SuppressWarnings;
+
 /**
  * A {@link PreferenceActivity} that presents a set of application settings. This
  * is the place where your profile is changed.
  */
+@SuppressFBWarnings(value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
+//Because of the structure with the static methods, a static attribute has to be set.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Can you elaborate on why we need the structure with static methods?

Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: This is the default usage of the settings screen. The only place where the static method is important is at the on preference change listener. This is one method which is triggered for all changes in preferences from the settings screen. Because this method uses the servicelocator, we have to set that in the onCreate.

I'm actually thinking that maybe we can statically get it from the NervousFish class?

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: > I'm actually thinking that maybe we can statically get it from the NervousFish class?

That is for certain true 😄 Also (if it still applies) can you add - a shorter version of - the description you gave here in the code 😇
